### PR TITLE
Enhancement/skale 4334 solidity 8

### PR DESCRIPTION
**Pull request by DimaStebaev**
Update to solidity 0.8.
Refactoring
Bug fixing.

### CODE REVIEW

File: proxy/contracts/MessageProxy.sol [hunk 1]
```diff
@@ -0,0 +1,347 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+/**
+ *   MessageProxyForMainnet.sol - SKALE Interchain Messaging Agent
```
Comment (payvint) [inline proxy/contracts/MessageProxy.sol]: MessageProxy instead of old name

File: proxy/contracts/mainnet/CommunityPool.sol [hunk 2]
```diff
@@ -97,28 +114,16 @@ contract CommunityPool is Twin {
                 Messages.encodeLockUserMessage(msg.sender)
             );
         }
-        msg.sender.transfer(amount);
+        payable(msg.sender).transfer(amount);
     }
 
     function setMinTransactionGas(uint newMinTransactionGas) external {
         require(hasRole(CONSTANT_SETTER_ROLE, msg.sender), "CONSTANT_SETTER_ROLE is required");
+        emit MinTransactionGasWasChanged(minTransactionGas, newMinTransactionGas);
```
Comment (payvint) [inline proxy/contracts/mainnet/CommunityPool.sol:122]: Is it good to emit event not at the end of execution?

File: proxy/contracts/mainnet/CommunityPool.sol [hunk 3]
```diff
@@ -35,11 +33,30 @@ import "./Linker.sol";
...
+        uint oldValue,
+        uint newValue
+    );
+
+    function initialize(
+        IContractManager contractManagerOfSkaleManagerValue,
+        Linker linker,
+        MessageProxyForMainnet messageProxyValue
+    )
+        external
```
Comment (payvint) [inline proxy/contracts/mainnet/CommunityPool.sol:53]: not public?

File: proxy/contracts/mainnet/DepositBoxes/DepositBoxERC1155.sol [hunk 4]
```diff
@@ -19,18 +19,20 @@
...
+import "@openzeppelin/contracts-upgradeable/token/ERC1155/extensions/IERC1155MetadataURIUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC1155/utils/ERC1155ReceiverUpgradeable.sol";
 import "../DepositBox.sol";
 import "../../Messages.sol";
 
 
 // This contract runs on the main net and accepts deposits
 contract DepositBoxERC1155 is DepositBox, ERC1155ReceiverUpgradeable {
 
+    using AddressUpgradeable for address;
```
Comment (payvint) [inline proxy/contracts/mainnet/DepositBoxes/DepositBoxERC1155.sol:34]: Maybe move it to DepositBox?

File: proxy/contracts/schain/TokenManager.sol [hunk 5]
```diff
@@ -188,5 +196,7 @@ abstract contract TokenManager is AccessControlUpgradeable, IMessageReceiver {
         tokenManagerLinker = newIMALinker;
         communityLocker = newCommunityLocker;        
         depositBox = newDepositBox;
+
+        emit DepositBoxWasChanged(address(0), newDepositBox);
```
Comment (payvint) [inline proxy/contracts/schain/TokenManager.sol:200]: Maybe let's add different event?

File: proxy/contracts/schain/TokenManagers/TokenManagerERC1155.sol [hunk 6]
```diff
@@ -37,6 +37,7 @@ import "../TokenManager.sol";
  * burns tokens.
  */
 contract TokenManagerERC1155 is TokenManager {
+    using AddressUpgradeable for address;
```
Comment (payvint) [inline proxy/contracts/schain/TokenManagers/TokenManagerERC1155.sol:40]: Add it to TokenManager?

File: proxy/contracts/test/MessageProxyForSchainTester.sol [hunk 7]
```diff
@@ -19,14 +19,13 @@
...
-pragma solidity 0.6.12;
-pragma experimental ABIEncoderV2;
+pragma solidity 0.8.6;
 
 import "../schain/MessageProxyForSchain.sol";
 
 contract MessageProxyForSchainTester is MessageProxyForSchain {    
 
-    constructor(KeyStorage _keyStorage, string memory schainName) public {
+    constructor(KeyStorage _keyStorage, string memory schainName) {
```
Comment (payvint) [inline proxy/contracts/test/MessageProxyForSchainTester.sol:28]: Somewhere we remove public, somewhere we add public in constructor
¯\_(ツ)_/¯

File: proxy/test/TokenManagerERC20.ts [hunk 8]
```diff
@@ -153,7 +154,7 @@ describe("TokenManagerERC20", () => {
 
     it("should reject with `Insufficient funds` if token balance is too low", async () => {
         // preparation
-        const error = "Insufficient funds";
+        const error = "insufficient funds";
```
Comment (payvint) [inline proxy/test/TokenManagerERC20.ts:157]: Looks like it would not pass

**Review by payvint — CHANGES_REQUESTED** [review]

File: proxy/test/TokenManagerERC20.ts (hunk 8 above)
Comment (payvint) [reply proxy/test/TokenManagerERC20.ts:157]: Solidity revert message with Upper capital letter

**Review by yavrsky — DISMISSED** [review]

File: proxy/contracts/mainnet/CommunityPool.sol (hunk 2 above)
Comment (DimaStebaev) [reply proxy/contracts/mainnet/CommunityPool.sol:122]: There are no external calls after it. Why not?

File: proxy/contracts/mainnet/CommunityPool.sol (hunk 3 above)
Comment (DimaStebaev) [reply proxy/contracts/mainnet/CommunityPool.sol:53]: It is not used from inside of smart contract. Should be external.

File: proxy/contracts/schain/TokenManager.sol (hunk 5 above)
Comment (DimaStebaev) [reply proxy/contracts/schain/TokenManager.sol:200]: Do you think that this event does not describe what happen?

File: proxy/contracts/test/MessageProxyForSchainTester.sol (hunk 7 above)
Comment (DimaStebaev) [reply proxy/contracts/test/MessageProxyForSchainTester.sol:28]: It's solidity 8 stuff: no public for constructors.

File: proxy/test/TokenManagerERC20.ts (hunk 8 above)
Comment (DimaStebaev) [reply proxy/test/TokenManagerERC20.ts:157]: It passes. (-:
I'm not sure why but more detailed error message are raised here. To match it I had to change capitalization.

File: proxy/contracts/MessageProxy.sol (hunk 1 above)
Comment (DimaStebaev) [reply proxy/contracts/MessageProxy.sol]: Done.

File: proxy/contracts/mainnet/DepositBoxes/DepositBoxERC1155.sol (hunk 4 above)
Comment (DimaStebaev) [reply proxy/contracts/mainnet/DepositBoxes/DepositBoxERC1155.sol:34]: Using statement is not inherited.

File: proxy/contracts/schain/TokenManagers/TokenManagerERC1155.sol (hunk 6 above)
Comment (DimaStebaev) [reply proxy/contracts/schain/TokenManagers/TokenManagerERC1155.sol:40]: Using statement is not inherited.

File: proxy/contracts/mainnet/DepositBoxes/DepositBoxERC1155.sol (hunk 4 above)
Comment (payvint) [reply proxy/contracts/mainnet/DepositBoxes/DepositBoxERC1155.sol:34]: Hmmm so we actively using `isContract` method from Address library everywhere(MessageProxy, DepositBoxes, TokenManagers) - without declaration using above

File: proxy/contracts/mainnet/CommunityPool.sol (hunk 3 above)
Comment (payvint) [reply proxy/contracts/mainnet/CommunityPool.sol:53]: Looks like we need to make every initialize external...

File: proxy/contracts/mainnet/CommunityPool.sol (hunk 2 above)
Comment (payvint) [reply proxy/contracts/mainnet/CommunityPool.sol:122]: I thought it is just code style guide

File: proxy/contracts/schain/TokenManager.sol (hunk 5 above)
Comment (payvint) [reply proxy/contracts/schain/TokenManager.sol:200]: We emit this event in initialize function

File: proxy/contracts/test/MessageProxyForSchainTester.sol (hunk 7 above)
Comment (payvint) [reply proxy/contracts/test/MessageProxyForSchainTester.sol:28]: Got it

File: proxy/test/TokenManagerERC20.ts (hunk 8 above)
Comment (payvint) [reply proxy/test/TokenManagerERC20.ts:157]: Interesting

**Review by payvint — APPROVED** [review]

**Pull request merged**